Added init sequence

yarondbb opened the pull request:
Added init sequence
Added rspec tests for Params module
Modified tests in params_test.rb to allign with changes

yarondbb committed:
Added init sequence
Added rspec tests for Params module
Modified tests in params_test.rb to allign with changes

yarondbb committed:
in process of writing documentation

yarondbb committed:
Added documentation.
Refactor params.init

kolmanv commented on test/params/params_spec.rb:
```diff
@@ -0,0 +1,240 @@
+      describe 'Params::parameter' do
+        it 'should raise error when parameter is defined as nil' do
+          Params.params_initialized = false
+          expect { Params::parameter('tmp', nil, 'tmp nil def') }.should raise_error \
```
This is not good. In many cases I want parameter as nil and in specific runs I will set it to some path or string or number.
The nil value let me do:
copy(Params.path) unless Prams.path  # If Params.to is nil, copy will not execute.

kolmanv replied:
The meaning of the value is default value which can be nil.

yarondbb replied:
The reason is to a simple type defiition and check.
The override system should check the types. and nil allows override with any value type.

One solution is to define a string value as '' and ask:  copy(Params.path) unless Prams.path.empty?
Other solution is to have the type added by the user:
Params.parameter my_param, nil, 'desc', 'String'

Also, I do not like parameters defined with no value, I think it might lead to errors.

kolmanv commented on test/params/params_spec.rb:
```diff
@@ -0,0 +1,240 @@
+    module Spec
+      describe 'Params::parameter' do
+        it 'should raise error when parameter is defined as nil' do
+          Params.params_initialized = false
```
Why this line is needed?

yarondbb replied:
all tests are added with this line to make the test portable.
Since this flag might be changed by a different test I want to make sure it is in its false mode.
There are other solutions such as make sure that if other test sets it to rtue (like the init tests then it would also set it back to false. To use this approach?

kolmanv replied:
This is good.

kolmanv commented on test/params/params_spec.rb:
```diff
@@ -0,0 +1,240 @@
+          Params::parameter('tmp4str', 'string_value', 'tmp4 def')
+          expect { Params::read_yml_params StringIO.new 'tmp4str: strr' }.should_not raise_error
+          expect { Params::read_yml_params StringIO.new 'tmp4str: 4' }.should raise_error \
+              "loaded yml param:'tmp4str' has different value type:'Fixnum'" + \
```
This is problematic because if I want a a string with value '4' it is a valid string.

kolmanv replied:
Maybe make it a warning but allow this.

yarondbb replied:
Ok I will make it same as the command parsing which cast all types to string is this is the defined type

kolmanv commented on test/params/params_spec.rb:
```diff
@@ -0,0 +1,240 @@
+          expect { Params::read_yml_params StringIO.new 'tmp4true: true' }.should_not raise_error
+          expect { Params::read_yml_params StringIO.new 'tmp4true: false' }.should_not raise_error
+
+          # false to other
```
You mean boolean to other?

yarondbb replied:
there is no boolean in ruby :) only TrueClass or FalseClass

kolmanv replied:
Still we may use boolean because all know what it means and inside we use TrueClass or FalseClass.

kolmanv commented on test/params/params_spec.rb:
```diff
@@ -0,0 +1,240 @@
+          expect { Params::parameter('tmp2', 'dummy', 'tmp2 not ok') }.should raise_error \
+              "parameters already initialized. No new definitions are allowed."
+        end
+      end
```
Please add tests:
1) When defining a parameter twice raise exception.

Bad:
Params.parameter('ku', 1, 'ku desk')
Params.parameter('ku', 2, 'ku desk2')

Good:
Params.parameter('ku', 1, 'ku desk')
Parameter.ku = 4  # change the default value.

yarondbb replied:
Ok I will change a bit the desing to do that

kolmanv commented on test/params/params_spec.rb:
```diff
@@ -0,0 +1,240 @@
+          expect { Params::parse_command_line_arguments ['--tmp7=true'] }.should raise_error
+          expect { Params::parse_command_line_arguments ['--tmp7=false'] }.should raise_error
+
+          # from TrueClass to other types.
```
There is no boolean? :)

yarondbb replied:
no.. TestClass or FalseClass

kolmanv commented on test/params/params_spec.rb:
```diff
@@ -0,0 +1,240 @@
+        it 'init should override parameters with file only' do
+          Params.params_initialized = false
+          Params::parameter('init_param', 'code', '')
+          File.stub(:exist?).and_return true
```
Very nice use of stubs.

kolmanv commented on test/params/params_spec.rb:
```diff
@@ -0,0 +1,240 @@
+          Params.init_param.should eq 'code'
+        end
+
+        it 'init should raise error when defining new param after init' do
```
Why is this nescessary?

yarondbb replied:
Exactly to address your comment before reagring re-definition.
Your advice is better:
1. allow only one call to Param.parameter for a given parameter)
2. let the file and command line override use the set instance param and not use Param.parameter.

I will implement as you suggest

kolmanv replied:
ok

kolmanv commented on lib/params.rb:
```diff
@@ -1,38 +1,173 @@
+# A project parameters module.
+# 1. Parameter definition (lowest precedence) - requires phases:
+#    e.g. Params.parameter '<param_name>', <param value>, '<description>'
+#    <param value> must not be nil (since it sets teh type of the parameter).
```
teh => the
I would not constraint this.
Better make 3 definition functions:
Parameter.int('an_integer_flag', 33, 'int description')
Parameter.string('a_string_flag', '33', 'string description')
Parameter.float('a_float_flag', 33.33, 'float description')

# Those not mandatory to implement now:

Parameter.array('an_array_flag', [33, 321, 1], 'array description')
Parameter.hash('a_hash_flag', 33, 'hash description')

yarondbb replied:
Let's talk about the nil contraint

kolmanv commented on lib/params.rb:
```diff
@@ -1,38 +1,173 @@
+#    Note that only defined parameters can be overridden. If new parameters are passed
+#    through file\command line, then an error will be raised.
+#    2.1 Input config file (medium precedence).
+#        Default configuration input file path is: '~/.bbfs/conf/<executable name>.conf'
```
I would consider setting the default value here to 'nil', usually I will run with command line arguments.
Better ask Alexey, our tester.

yarondbb replied:
If file is not there it would not be used.
I think it is good to set a defualt value to this parameter (help to orginaize the execution in our dirs structure) under .bbgs

kolmanv replied:
Problem will happen when a person have the config file (which is very likely) but does not intend to use it. Then for passive behavior of not using the file he will have to do an active behavior of setting the flag to empty string or nil. This is counter intuitive better to use it if he specified in the flag to use it.
Yes it helps to organize the dirs but I think we should have a empty string/nil value here anyway.

kolmanv commented on lib/params.rb:
```diff
@@ -1,38 +1,173 @@
+    end
+
+
+    def Params.print_all_project_parameters
```
print_all_project_parameters => print_all_parameters
This is not all project parameters, but only part because some files are not included/required by all executables.

yarondbb replied:
right. mybe use: print_global_params

kolmanv replied:
?print_params

kolmanv commented on lib/params.rb:
```diff
@@ -1,38 +1,173 @@
+      counter=0
+      names.each do |name|
+        tmp_name = name.to_s
+        tmp_name.slice!(0)
```
What does this do?

yarondbb replied:
opps forgot documenting..

kolmanv commented on lib/params.rb:
```diff
@@ -1,38 +1,173 @@
+
+    # Initializes the project parameters
+    def Params.init args
+      Params.parameter 'conf_file', \
```
I would start with parse command line arguments first.
Then check if yml file is overridden or not, actually you don't have to, just check Params.conf_file.
The downside here is that you cannot override some parameters now.
You can help yourself by adding auxiliary data to each parameter.
But I don't like the regexp you use to find out whether the config_file was overriden.
Think what is simpler. and easy to maintain.

yarondbb replied:
Ok. I liked your suggestion to use the parser and save the param=value it creates, then check for file override and use it, then use the command line parser data to override again

kolmanv replied:
You will have to change Alexey parse function a bit for that.

kolmanv commented on lib/params.rb:
```diff
@@ -1,38 +1,173 @@
+    def Params.override_config_file_path args
+      args.each do |argument|
+        # check parameter format: '--conf_file=<file_path>'
+        /--conf_file=([a-zA-Z0-9\-_\.]+)/ =~ argument
```
You have a parsing parameters function, use it.
Make it return a hash and not write itself directly to the Params module.
Then after config file checks and loads merge the hash to Params module to sustain precedence.

yarondbb replied:
Ok. good approach. will use it

kolmanv commented on lib/params.rb:
```diff
@@ -1,38 +1,173 @@
+    # Default value can not be nil.
+    def Params.parameter(name, value, description)
+      if value.nil?
+        raise "parameter:'#{name}' value can not be nil."
```
I think we should allow nil values.

yarondbb replied:
Let's talk about it

kolmanv commented on lib/params.rb:
```diff
@@ -1,38 +1,173 @@
+          elsif new_value_upcase.eql? 'FALSE'
+            self.instance_variable_set '@' + name, false
+          else
+            raise "Defined value type is boolean but overriding value is not boolean."
```
Write what are the options, i.e., 'true' or 'false' only
 or add more 't' 'f' 'y' 'n' 'yes' 'no' ......

yarondbb replied:
ok

kolmanv commented on lib/params.rb:
```diff
@@ -1,38 +1,173 @@
+              #true\false case
+              Params.parameter param_name, proj_params[param_name], ''
+            else
+              #not cast from fixnum to float and not true\false case
```
fixnum and int are almost the same if we use them as parameters.
consider leaving only one.

yarondbb replied:
I will look into it

kolmanv replied:
I mean we never change the parameters once they are defined.

kolmanv commented:
Good work. If you 100% sure how to fix everything please merge after you fix.

yarondbb commented:
Ok. I will talk with you if not 100% sure of anything (like the nil and fixnum\int and boolean)

kolmanv commented on lib/params.rb:
```diff
@@ -1,38 +1,173 @@
+      existing_value = self.instance_variable_get '@' + name
+      if existing_value.nil?
+        # new parameter
+        self.instance_variable_set '@' + name, value
```
I'm thinking of changing this approach, while it is comfortable to use it seems kinda wrong because we have more and more member methods and functions, i.e., names which parameters cannot be like I cannot make a parameter called parameter:
Params.parameter('parameter', 3, 'desc')
Because it will override the definition function (parameter method).
I think of using simple hash and set and get methods, maybe with []
Params[:parameter]  # with symbols
Params['file_path']  # with strings
Don't know what better.

yarondbb committed:
added tests for string type

yarondbb committed:
Changed param design to use access operator [].
Changed param design to use explicit type methods for definition
Added example file

yarondbb committed:
Changed Params usage from Parmas.<name> to Params['<name>'] for the entire project

yarondbb committed:
Changed some comments

yarondbb committed:
Added init sequence to bin files
